xcube/webapi/ows/coverages/controllers.py

```diff
@@ -144,34 +194,249 @@ def get_coverage_data(
                 [type_ for value in media_types.values() for type_ in value]
             )
         )
+    final_bbox = get_bbox_from_ds(ds)
+    if not is_xy_order(final_crs):
+        final_bbox = final_bbox[1], final_bbox[0], final_bbox[3], final_bbox[2]
+    return content, final_bbox, final_crs
+
+
+def _assert_coverage_size_ok(ds: xr.Dataset, scale_factor: float):
+    size_limit = 4000 * 4000  # TODO make this configurable
+    h_dim = _get_h_dim(ds)
+    v_dim = _get_v_dim(ds)
+    for d in h_dim, v_dim:
+        size = ds.dims[d]
+        if size == 0:
+            # Requirement 8C currently specifies a 204 rather than 404 here,
+            # but spec will soon be updated to allow 404 as an alternative.
+            # (J. Jacovella-St-Louis, pers. comm., 2023-11-27).
+            raise ApiError.NotFound(
+                f'Requested coverage contains no data: {d} has zero size.'
+            )
+    if (h_size := ds.dims[h_dim] / scale_factor) * (
+        y_size := ds.dims[v_dim] / scale_factor
+    ) > size_limit:
+        raise ApiError.ContentTooLarge(
+            f'Requested coverage is too large:'
+            f'{h_size} × {y_size} > {size_limit}.'
+        )
 
 
 _IndexerTuple = NamedTuple(
-    'Indexers', [('indices', dict[str, Any]), ('slices', dict[str, slice])]
+    'Indexers',
+    [
+        ('indices', dict[str, Any]),  # non-geographic single-valued specifiers
+        ('slices', dict[str, slice]),  # non-geographic range specifiers
+        ('x', Optional[Union[float, tuple[float, float]]]),  # x or longitude
+        ('y', Optional[Union[float, tuple[float, float]]]),  # y or latitude
+    ],
 )
 
 
-def _subset_to_indexers(subset_spec: str, ds: xr.Dataset) -> _IndexerTuple:
-    indices, slices = {}, {}
-    for part in subset_spec.split(','):
-        axis, low, high = re.match(
-            '^(.*)[(]([^:)]*)(?::(.*))?[)]$', part
-        ).groups()
-        if high is None:
-            # TODO: parse to float if possible
-            indices[axis] = low
-        else:
-            # TODO: don't flip longitude
-            if (low < high) != (ds[axis][0] < ds[axis][-1]):
-                low, high = high, low
-            if axis != 'time':
-                # TODO: catch parse exception and pass through as string
+def _apply_subsetting(
```

Comment:
Mmmaybe. But it's currently quite tightly coupled to the OGC API - Coverages model -- amongst other things, this means that the specifier mixes non-spatial specifiers (which refer directly to a native dataset dimension) and spatial specifiers (which define a bounding box or other area in an arbitrary CRS which may have quite a complex transformation to the nearest equivalent subset in the native dataset). We have to implement this for conformance, but I'm not convinced by either the conceptual elegance or the general usefulness of the approach (and not sure that OGC are either, since they also mandate a bbox parameter!).

I suggest that we wait for the planned unification of the WCS and OGC API - Coverages implementations, which will naturally lead to code consolidation. Whatever we consolidate there we can also consider moving into the core.